### ISSUE TITLE
False positive in unreachable_from_main

### ISSUE THREAD

**Issue by natebosch** [dart-lang/linter#4495]
`jsonEncode` checks for the presence of a `toJson` method and uses it when it exists. It's a common pattern to have a `toJson` without any static references to it. I think we should consider it a false positive to flag a `toJson` member.

```dart
import 'dart:convert';

void main() async {
  print(jsonEncode([C()]));
}

class C {
  List toJson() => ['c'];
  //   ^ unreachable_from_main
}
```

**Comment by srawlins** [comment]
I'm happy to take this as a false positive, but I am _extremely_ curious: how does `dart:convert#jsonEncode` do this? Mirrors? `native` magic?

**Comment by natebosch** [comment]
> I am _extremely_ curious: how does `dart:convert#jsonEncode` do this?

Cast the object to dynamic for dynamic dispatch, use `try {} catch {}` for the `NoSuchMethodError` when it doesn't exist.

**Comment by srawlins** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/954ed42cf8988657ed4faf2aed9459be87588486
